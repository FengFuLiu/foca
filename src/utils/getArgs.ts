const slice = Array.prototype.slice;

export const getArgs = <T = any>(args: IArguments): T[] => slice.call(args);
