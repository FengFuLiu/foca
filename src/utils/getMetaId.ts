export const getMetaId = (id: number | string) => '_' + id;
