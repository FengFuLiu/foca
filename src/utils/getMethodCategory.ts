export const getMethodCategory = (methodName: string) =>
  methodName.indexOf('_') === 0 ? 'hidden' : 'visible';
