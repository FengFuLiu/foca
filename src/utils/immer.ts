import { Immer, enableES5 } from 'immer';

/**
 * 支持ES5
 * @link https://immerjs.github.io/immer/docs/installation#pick-your-immer-version
 * @since immer 6.0
 */
enableES5();

export const immer = new Immer({
  autoFreeze: false,
});
