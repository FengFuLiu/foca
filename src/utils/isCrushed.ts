/*
 * The browsers which don't support Function.name should be considered as compressed.
 *
 * @link https://github.com/terser/terser
 * toplevel (default false) - set to true if you wish to enable top level variable
 * and function name mangling and to drop unused variables and functions.
 *
 * For ESM, don't put CustomFunction to toplevel.
 */
const name = (() => {
  function CustomFunction() {}
  return CustomFunction.name;
})();

const crushed = typeof name !== 'string' || name !== 'CustomFunction';

export const isCrushed = () => crushed;
