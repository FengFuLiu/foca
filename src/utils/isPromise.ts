export const isPromise = <T>(value: any): value is Promise<T> => {
  return (
    value instanceof Promise ||
    (value !== null &&
      (typeof value === 'object' || typeof value === 'function') &&
      typeof value.then === 'function')
  );
};
