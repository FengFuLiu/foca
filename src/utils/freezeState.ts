import { freeze } from 'immer';

export const freezeState = <T = any>(state: T): T => {
  return process.env.NODE_ENV === 'production' ? state : freeze(state, true);
};
