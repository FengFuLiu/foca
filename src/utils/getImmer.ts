import { Immer, enableES5 } from 'immer';

const immer = new Immer({
  autoFreeze: false,
});

/**
 * support for the fallback implementation has to be explicitly enabled
 * @link https://immerjs.github.io/immer/docs/installation#pick-your-immer-version
 * @since immer 6.0
 */
enableES5();

export const getImmer = () => immer;
