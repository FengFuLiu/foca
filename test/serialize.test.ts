import { parseState, stringifyState } from '../src/utils/serialize';

it('can clone basic data', () => {
  expect(
    parseState(
      stringifyState({
        x: 1,
        y: 'y',
        z: true,
      }),
    ),
  ).toMatchObject({
    x: 1,
    y: 'y',
    z: true,
  });
});

it('can clone complex data', () => {
  expect(
    parseState(
      stringifyState({
        x: 1,
        y: {
          z: [1, 2, '3'],
        },
      }),
    ),
  ).toMatchObject({
    x: 1,
    y: {
      z: [1, 2, '3'],
    },
  });
});

it('can clone null and undefined', () => {
  const data = {
    x: [
      undefined,
      1,
      {
        test: undefined,
        test1: 'hello',
      },
      null,
      undefined,
    ],
    y: null,
    z: undefined,
  };

  expect(stringifyState(data)).toMatchSnapshot();
  expect(parseState(stringifyState(data))).toMatchSnapshot();
  expect(parseState(stringifyState(data))).toMatchObject(data);
});
