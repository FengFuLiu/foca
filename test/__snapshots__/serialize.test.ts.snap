// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can clone null and undefined 1`] = `"{\\"x\\":[\\"__JSON_UNDEFINED__\\",1,{\\"test\\":\\"__JSON_UNDEFINED__\\",\\"test1\\":\\"hello\\"},null,\\"__JSON_UNDEFINED__\\"],\\"y\\":null,\\"z\\":\\"__JSON_UNDEFINED__\\"}"`;

exports[`can clone null and undefined 2`] = `
Object {
  "x": Array [
    undefined,
    1,
    Object {
      "test": undefined,
      "test1": "hello",
    },
    null,
    undefined,
  ],
  "y": null,
  "z": undefined,
}
`;
